nixos/modules/services/misc/matrix-synapse.nix

```diff
@@ -176,548 +169,497 @@ in {
+                if you like.  Any required intermediary certificates can be
+                appended after the primary certificate in hierarchical order.
+              '';
+            };
+
+            tls_private_key_path = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "/var/lib/acme/example.com/key.pem";
+              description = ''
+                PEM encoded private key for TLS. Specify null if synapse is not
+                speaking TLS directly.
+              '';
+            };
+
+            presence.enabled = mkOption {
+              type = types.bool;
+              default = false;
+              example = true;
+              description = ''
+                Whether to enable presence tracking.
+
+                Presence tracking allows users to see the state (e.g online/offline)
+                of other local and remote users.
+              '';
+            };
+
+            listeners = mkOption {
               type = types.listOf (types.submodule {
                 options = {
-                  names = mkOption {
+                  port = mkOption {
+                    type = types.port;
+                    example = 8448;
+                    description = ''
+                      The port to listen for HTTP(S) requests on.
+                    '';
+                  };
+
+                  bind_addresses = mkOption {
                     type = types.listOf types.str;
+                    default = [
+                      "::1"
+                      "127.0.0.1"
+                    ];
+                    example = [
+                      "::"
+                      "0.0.0.0"
+                    ];
                     description = ''
-                      List of resources to host on this listener.
+                     IP addresses to bind the listener to.
                     '';
-                    example = ["client" "federation"];
                   };
-                  compress = mkOption {
+
+                  type = mkOption {
+                    type = types.enum [
+                      "http"
+                      "manhole"
+                      "metrics"
+                      "replication"
+                    ];
+                    default = "http";
+                    example = "metrics";
+                    description = ''
+                      The type of the listener, usually http.
+                    '';
+                  };
+
+                  tls = mkOption {
+                    type = types.bool;
+                    default = true;
+                    example = false;
+                    description = ''
+                      Whether to enable TLS on the listener socket.
+                    '';
+                  };
+
+                  x_forwarded = mkOption {
                     type = types.bool;
+                    default = false;
```

Comment:
Does it make sense to set this to true if tls is false? In my mind I cant imagine a situation where you'd run with tls off exposed without trusting you environment not to send fake X_FORWARDED headers

Reply:
People who set this should be highly aware, that they allow anyone, who can reach synapse directly, to spoof IP addresses. I'm hesistant to include implicit behaviour when it comes with strings like these attached.